### ISSUE TITLE
InvalidCastException when trying to set entityimage of systemuser entity (Duplicate #154 & #153)

### ISSUE THREAD

**Issue by mwolfaardt** [seanmcne/Microsoft.Xrm.Data.PowerShell#157]
When trying to execute the following : 

$users = Get-CrmRecords -EntityLogicalName systemuser -FilterAttribute domainname -FilterOperator eq -FilterValue "LMiller@d365trial.onmicrosoft.com" -Fields systemuserid,fullname,entityimage

    $picture = [System.IO.File]::ReadAllBytes("c:\img\LMiller.jpg")

$users.CrmRecords | 
foreach{
$user = $_
$user.entityimage = $picture
#$user.sample_name = "60.jpg"
set-CrmRecord $user
}

I get an exception 

Specified cast is not valid.
At C:\Users\user1\Documents\WindowsPowerShell\Modules\Microsoft.Xrm.Data.Powershell\2.4\Microsoft.Xrm.Data.Powershell.psm1:772 char:9
+         throw $conn.LastCrmException
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (:) [], InvalidCastException
    + FullyQualifiedErrorId : Specified cast is not valid.

**Comment by seanmcne** [comment]
@mwolfaardt The issue may be that you're not encoding the string at all - when you readAllBytes your return type should be a byte array and not a string which is probably why you're getting an invalid cast.  Try this instead: 

$picture = [System.IO.File]::ReadAllBytes("c:\img\LMiller.jpg")
$base64String =[System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($picture))
$users.CrmRecords |
  foreach{
    $user = $_
    $user.entityimage = $base64String
    #$user.sample_name = "60.jpg"
    set-CrmRecord $user
}

**Comment by mwolfaardt** [comment]
Thank you
I tried what the solution, but now I'm getting Type Mismatch: Type of Attribute: SystemUser.entityimage is: System.Byte[]. However, Type of passed-in value is: System.String

**Comment by seanmcne** [comment]
You're right I'm seeing the type actuall is a byte array. I'll have to try and debug this and test it out, not sure why it isn't working with your first script at this point but when I get a chance I'll debug the issue. Unless Ken can beat me to it 🙂

**Comment by mwolfaardt** [comment]
Thank you, I'm no expert but it seems in the powershell script it checks what data type the field is, and then somewhere a value gets set to null

**Comment by mwolfaardt** [comment]
My purpose is really simple, take a picture and set the user image to the picture. I would be great if CRM used the user profile picture from exchange. I've also got powershell for that if you would like to add it to your examples, sync exchange profile picture with crm user picture.  =)

**Comment by mwolfaardt** [comment]
any update on this issue?

**Comment by seanmcne** [comment]
@mwolfaardt I haven't had a chance to dig into this yet due to some other commitments consuming my time right at the moment.  This is also an open community project with open source - if you know where the issue is or have debugged it please let us know (or if you know the specific fix you can also submit a pull request with the fix).  I'll also tag this for "Help Wanted" in addition.  Thanks!

**Comment by seanmcne** [comment]
@mwolfaardt I just took a stab at this on my end and everything worked fine without any errors on my end.  Here's the script: 

$picture = [System.IO.File]::ReadAllBytes("c:\img\LMiller.jpg") #60px jpg image 
Set-CrmRecord -EntityLogicalName systemuser -id "53ad51fd-65b6-e611-80f2-5065f38bb4d1" @{"entityimage"=$picture} -Verbose

after running this with a test image, I went into CRM and now see that this user has that image as their entity picture.  Can you try using the above method of updating the user and see if it works on your end?

**Comment by mwolfaardt** [comment]
@seanmcne Thank you for the feedback. I've updated my powershell script to use your code. I've tried to execute the code with a systemuser that has no image as well as one that has. I've also tried with different images, and I still get the same error message. 

This is more exception details : 

LastCrmError                   : Specified cast is not valid.
LastCrmException               : System.InvalidCastException: Specified cast is not valid.
                                    at Microsoft.Xrm.Tooling.Connector.CrmServiceClient.AddValueToPropertyList(KeyValuePair`2 Field, AttributeCollection 
                                 PropertyList)
                                    at Microsoft.Xrm.Tooling.Connector.CrmServiceClient.UpdateEntity(String entityName, String keyFieldName, Guid id, 
                                 Dictionary`2 fieldList, String applyToSolution, Boolean enabledDuplicateDetection, Guid batchId)
                                    at CallSite.Target(Closure , CallSite , Object , Object , Object , Object , Object , Object , Boolean , Guid )

**Comment by mwolfaardt** [comment]
I'm also using Dynamics 365 CRM Online, don't know if that makes any difference. 
ConnectedOrgVersion : 8.2.0.720

**Comment by seanmcne** [comment]
Strange let me try again when I get a chance and see if It works again. If it still works for me I'll share the whole script with you in addition to the image to make sure we are trying the same thing.

**Comment by mwolfaardt** [comment]
Thank you

**Comment by amervitz** [comment]
I'm able to reproduce this error when creating a new record as in @seanmcne's sample code.

The problem is the type assigned to CrmDataTypeWrapper object is CrmBoolean but it should be Raw:

> PS> $newfields.entityimage.Type
> PS> CrmBoolean

Adding a default case to the switch statement in the code above to assign $newfield.Type to be Raw should fix it in this case.

It works for me though when updating an existing record through Set-CrmRecord -CrmRecord and when the record already has a value in the entityimage field.

I can submit a PR for the case I've found if you like.

**Comment by mwolfaardt** [comment]
Yes can you submit a PR

**Comment by seanmcne** [comment]
Ok I now can repro, the method I used to update successfully was updating a user that already had an image and changing it.  @amervitz  if you want to create a quick PR go ahead, otherwise I can try to do this later today or early next week (I'm on vacation and should have some time to look more at a few of these issues/bugs).

**Comment by amervitz** [comment]
@seanmcne I'll send a PR shortly.

**Comment by amervitz** [comment]
This issue is actually already addressed in the master branch, the problematic code was in Set-CrmRecord in the 2.4 release, but it's now in the newly added MapFieldTypeByFieldValue function which assigns the correct CrmFieldType by default.

**Comment by seanmcne** [comment]
Ahh of course, this is the same issue that was fixed in #154 and #153 

I'll go ahead and close this then as our next release will fix the issue. Thanks @amervitz!

**Comment by mwolfaardt** [comment]
Thank you everyone that assisted on this :)
